### PR TITLE
kubelet: fix e2e-node cgroups test on cgroup v2

### PR DESCRIPTION
**Pull request by giuseppe**
**What type of PR is this?**
/kind bug
**What this PR does / why we need it**:

adapt e2e node tests for running on a system with cgroup v2.

A couple of issues found while working on the tests are fixed as separate patches.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

[KEP]: https://github.com/kubernetes/enhancements/pull/1370

### CODE REVIEW

File: pkg/kubelet/cm/cgroup_manager_linux.go [hunk 1]
```diff
@@ -246,6 +255,13 @@ func updateSystemdCgroupInfo(cgroupConfig *libcontainerconfigs.Cgroup, cgroupNam
 
 // Exists checks if all subsystem cgroups already exist
 func (m *cgroupManagerImpl) Exists(name CgroupName) bool {
+	if libcontainercgroups.IsCgroup2UnifiedMode() {
```
Comment (derekwaynecarr) [inline pkg/kubelet/cm/cgroup_manager_linux.go:258]: this eliminates verification on particular controllers being present and available on a cgroupv2 host.

was that intentional?

File: pkg/kubelet/cm/cgroup_manager_linux.go [hunk 2]
```diff
@@ -388,6 +404,62 @@ func getCpuMax(cpuQuota *int64, cpuPeriod *uint64) string {
...
+	if err := os.MkdirAll(path, 0755); err != nil {
+		return fmt.Errorf("failed to create cgroup %q : %v", path, err)
+	}
+
+	// Retrieve all the supported controllers from the cgroup root
+	controllersFileContent, err := ioutil.ReadFile(filepath.Join(cmutil.CgroupRoot, "cgroup.controllers"))
+	if err != nil {
+		return fmt.Errorf("failed to read controllers from %q : %v", cmutil.CgroupRoot, err)
+	}
+	// The retrieved content looks like: "cpuset cpu io memory hugetlb pids".  Prepend each of the controllers
```
Comment (derekwaynecarr) [inline pkg/kubelet/cm/cgroup_manager_linux.go:418]: this appears to blindly take all the root controllers that were enabled on the system and propagates it to the children.  shouldn't we only propagate the controllers that are supported by k8s itself?  this feels like the place we should continue to maintain a whitelist of supported controllers.

File: test/e2e_node/services/kubelet.go [hunk 3]
```diff
@@ -194,6 +194,10 @@ func (e *E2EServices) startKubelet() (*server, error) {
 		unitName := fmt.Sprintf("kubelet-%s.service", unitTimestamp)
 		cmdArgs = append(cmdArgs,
 			systemdRun,
+			"-p", "Delegate=true",
```
Comment (derekwaynecarr) [inline test/e2e_node/services/kubelet.go:197]: thanks for updating this to ensure its explicit, i think we had previously assumed this was done globally via systemd config.

**Review by derekwaynecarr — CHANGES_REQUESTED** [review]
the test changes look good, i am curious about the controller propogation and if it needs whitelisting.

File: pkg/kubelet/cm/cgroup_manager_linux.go (hunk 2 above)
Comment (giuseppe) [reply pkg/kubelet/cm/cgroup_manager_linux.go:418]: good idea.  I will change the test to only enable the controllers we need.

Should `pids` be behind the feature gate?

File: pkg/kubelet/cm/cgroup_manager_linux.go (hunk 1 above)
Comment (giuseppe) [reply pkg/kubelet/cm/cgroup_manager_linux.go:258]: it was intentional as on cgroup v2 the cgroup existence depend only on the cgroup directory being created.  Should I change the test to also verify if the controllers we need are enabled?

Comment (giuseppe) [reply pkg/kubelet/cm/cgroup_manager_linux.go:259]: I've changed the implementation to also check if the controllers we need are enabled.

File: pkg/kubelet/cm/cgroup_manager_linux.go (hunk 2 above)
Comment (giuseppe) [reply pkg/kubelet/cm/cgroup_manager_linux.go:464]: thanks, done

**Comment by giuseppe** [comment]
it is ready for review

**Comment by giuseppe** [comment]
@derekwaynecarr can we move forward with this PR?

**Comment by derekwaynecarr** [comment]
thanks for the updates.

are you working to also get e2e running on a v2 aware host reporting back into the community?

/lgtm
/approve

**Comment by giuseppe** [comment]
> thanks for the updates.
> 
> are you working to also get e2e running on a v2 aware host reporting back into the community?

thanks for the review.

Not yet, but it is my next goal

**Pull request merged**